tests/tests/connectors/test-class-connector-gravityforms.php

```diff
@@ -0,0 +1,949 @@
+				'user_role' => 'admin',
+			)
+		);
+		\wp_set_current_user( $user_id );
+		$this->mock->expects( $this->once() )
+			->method( 'log' )
+			->with(
+				$this->equalTo( __( 'Note #%1$d added to lead #%2$d on "%3$s" form', 'stream' ) ),
+				$this->callback(
+					function( $args ) {
+						return ! empty( $args['form_title' ] ) && $args['form_title'] === 'Test form';
+					}
+				),
+				$this->greaterThan( 0 ),
+				$this->equalTo( 'notes' ),
+				$this->equalTo( 'added' )
+			);
+
+		// Do stuff.
+		$form_id  = $this->create_form();
+		$form     = \GFAPI::get_form( $form_id );
+		$entry_id = \GFAPI::add_entry(
+			array(
+				'id'               => '1',
+				'form_id'          => $form_id,
+				'date_created'     => '2016-03-22 19 => 13 => 19',
+				'is_starred'       => 0,
+				'is_read'          => 0,
+				'ip'               => '192.168.50.1',
+				'source_url'       => 'http => \/\/local.wordpress.dev\/?gf_page=preview&id=1',
+				'post_id'          => null,
+				'currency'         => 'USD',
+				'payment_status'   => null,
+				'payment_date'     => null,
+				'transaction_id'   => null,
+				'payment_amount'   => null,
+				'payment_method'   => null,
+				'is_fulfilled'     => null,
+				'created_by'       => '1',
+				'transaction_type' => null,
+				'user_agent'       => 'Mozilla\/5.0 (Macintosh; Intel Mac OS X 10_11_1) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/48.0.2564.116 Safari\/537.36',
+				'status'           => 'active',
+			)
+		);
+		$note    = 'Lorem ipsum dolor...';
+		\GFFormsModel::add_note( $entry_id, $user_id, 'johndoe', $note, 'user' );
+
+		$this->assertFalse( 0 === did_action( 'wp_stream_test_callback_gform_post_note_added' ) );
+	}
+
+	public function test_callback_gform_pre_note_deleted() {
+		global $wpdb;
+
+		$user_id = self::factory()->user->create(
+			array(
+				'username'  => 'johndoe',
+				'user_role' => 'admin',
+			)
+		);
+		\wp_set_current_user( $user_id );
+		$form_id  = $this->create_form();
+		$form     = \GFAPI::get_form( $form_id );
+		$entry_id = \GFAPI::add_entry(
+			array(
+				'id'               => '1',
+				'form_id'          => $form_id,
+				'date_created'     => '2016-03-22 19 => 13 => 19',
+				'is_starred'       => 0,
+				'is_read'          => 0,
+				'ip'               => '192.168.50.1',
+				'source_url'       => 'http => \/\/local.wordpress.dev\/?gf_page=preview&id=1',
+				'post_id'          => null,
+				'currency'         => 'USD',
+				'payment_status'   => null,
+				'payment_date'     => null,
+				'transaction_id'   => null,
+				'payment_amount'   => null,
+				'payment_method'   => null,
+				'is_fulfilled'     => null,
+				'created_by'       => '1',
+				'transaction_type' => null,
+				'user_agent'       => 'Mozilla\/5.0 (Macintosh; Intel Mac OS X 10_11_1) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/48.0.2564.116 Safari\/537.36',
+				'status'           => 'active',
+			)
+		);
+
+		// Wow, I actually had to do this just to get the Note ID.
```

Comment:
@kidunot89 can you explain more about this comment and complexity? Is this something we need to be aware?

Reply:
Because the function doesn't return the ID and retrieving the it from `$wpdb` fails.

So I have to use a nearby hook in the function's logic and pass the ID back by reference.

Reply:
Makes sense! 😀

Your comment had me curious and thanks for explaining!

Reply:
@kopepasah Later versions of the GravityForms have alternative to this that returns the Note ID.